xdsl/transforms/csl_stencil_handle_async_flow.py

```diff
@@ -0,0 +1,238 @@
+from dataclasses import dataclass
+
+from xdsl.builder import ImplicitBuilder
+from xdsl.context import MLContext
+from xdsl.dialects import arith, scf
+from xdsl.dialects.builtin import (
+    FunctionType,
+    IndexType,
+    IntegerAttr,
+    MemRefType,
+    ModuleOp,
+    SymbolRefAttr,
+)
+from xdsl.dialects.csl import csl, csl_stencil, csl_wrapper
+from xdsl.ir import (
+    Attribute,
+    Block,
+    Operation,
+    OpResult,
+    Region,
+)
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+from xdsl.utils.hints import isa
+
+
+@dataclass(frozen=True)
+class HandleCslStencilApplyAsyncCF(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_stencil.ApplyOp, rewriter: PatternRewriter, /):
+        if op.next_op and isinstance(op.next_op, csl.ReturnOp):
+            return
+        if (
+            op.next_op
+            and isinstance(call_op := op.next_op, csl.CallOp)
+            and op.next_op.next_op
+            and isinstance(op.next_op.next_op, csl.ReturnOp)
+        ):
+            terminator = op.post_process.block.last_op
+            assert isinstance(terminator, csl_stencil.YieldOp)
+            rewriter.insert_op(call_op.clone(), InsertPoint.before(terminator))
+            rewriter.erase_op(call_op)
```

Comment:
Does this still work if `call_op` had a result?

I'm also somewhat confused as to what this rewrite does in general

Reply:
There is an implicit assumption that the `call_op` was placed there by the other rewrite pass. It's not finished yet though.